### PR TITLE
Added support for using PAM with basic-auth. 

### PR DESCRIPTION
**Pull request by arimed**
@erictune If basic-auth is used for authentication when a client communicates with the api server
and the BASIC_AUTH_PAM env var is on, then the configured PAM module is used to validate the
user name/password pair.  For example the local PAM module may be configured to lookup the user name/password in Active Directory.

### CODE REVIEW

File: plugin/pkg/auth/authenticator/request/basicauth/basicauth.go [hunk 1]
```diff
@@ -33,11 +36,52 @@ func New(auth authenticator.Password) *Authenticator {
 	return &Authenticator{auth}
 }
 
+type Credentials struct {
```
Comment (liggitt) [inline plugin/pkg/auth/authenticator/request/basicauth/basicauth.go]: these belong in a separate file, implemented as a authenticator.Password. See packages under https://github.com/kubernetes/kubernetes/tree/master/plugin/pkg/auth/authenticator/password

File: hack/lib/golang.sh [hunk 2]
```diff
@@ -397,7 +397,7 @@ kube::golang::build_binaries_for_platform() {
         "${nonstatics[@]:+${nonstatics[@]}}"
     fi
     if [[ "${#statics[@]}" != 0 ]]; then
-      CGO_ENABLED=0 go install -installsuffix cgo "${goflags[@]:+${goflags[@]}}" \
+      CGO_ENABLED=1 go install -installsuffix cgo "${goflags[@]:+${goflags[@]}}" \
```
Comment (liggitt) [inline hack/lib/golang.sh]: not sure what the implications of this change are, but they might be big

**Comment by liggitt** [comment]
rather than triggering off envvars, the PAM impl should be a particular password authenticator that could be plugged into the basic auth request authenticator

File: plugin/pkg/auth/authenticator/request/basicauth/basicauth.go [hunk 3]
```diff
@@ -33,11 +36,52 @@ func New(auth authenticator.Password) *Authenticator {
...
+}
+
 // AuthenticateRequest authenticates the request using the "Authorization: Basic" header in the request
 func (a *Authenticator) AuthenticateRequest(req *http.Request) (user.Info, bool, error) {
 	username, password, found := req.BasicAuth()
 	if !found {
 		return nil, false, nil
 	}
+
+	if os.Getenv("BASIC_AUTH_PAM") == "1" {
```
Comment (liggitt) [inline plugin/pkg/auth/authenticator/request/basicauth/basicauth.go]: All of these changes belong in a PAMPassword impl and package

File: plugin/pkg/auth/authenticator/request/basicauth/basicauth.go [hunk 4]
```diff
@@ -17,10 +17,13 @@ limitations under the License.
 package basicauth
 
 import (
-	"net/http"
-
+	"errors"
+	"github.com/golang/glog"
+	"github.com/msteinert/pam"
```
Comment (liggitt) [inline plugin/pkg/auth/authenticator/request/basicauth/basicauth.go]: All packages have to be vendored in godeps. Add this package in a separate commit

File: hack/lib/golang.sh (hunk 2 above)
Comment (liggitt) [inline hack/lib/golang.sh]: I wouldn't expect to need this change. Assuming there are platform dependencies, add build tags to the PAM password impl to limit dependencies to the PAM lib to linux (or whatever other platforms support it) and stub the impl for other platforms

**Comment by arimed** [comment]
I just submitted a CLA for Centrify corp

File: hack/lib/golang.sh (hunk 2 above)
Comment (arimed) [inline hack/lib/golang.sh]: Pam uses a C lib, so CGO must be enabled.

Comment (liggitt) [inline hack/lib/golang.sh]: Doesn't building with cgo make it link to libraries dynamically, not statically? If so, that is reversing the intent of this part of the build script

Comment (liggitt) [inline hack/lib/golang.sh]: https://github.com/kubernetes/kubernetes/commit/9a901ed46db225f588cc691feeb47e7ace92bbc8 explicitly made kube-apiserver a static binary. If we want to reverse that to add a cgo dependency, we'll need broader agreement

Comment (arimed) [inline hack/lib/golang.sh]: without CGO_ENABLED=1, I get the following build error:
icauth/basicauth.go:22:2: C source files not allowed when not using cgo or SWIG: transaction.c

**Comment by erictune** [comment]
@lavalamp or @thockin to comment on the proposal here to enable CGO.

File: plugin/pkg/auth/authenticator/request/basicauth/basicauth.go [hunk 5]
```diff
@@ -33,11 +36,52 @@ func New(auth authenticator.Password) *Authenticator {
...
 	if !found {
 		return nil, false, nil
 	}
+
+	if os.Getenv("BASIC_AUTH_PAM") == "1" {
+		c := Credentials{
+			Password: password,
+		}
+
+		module := os.Getenv("KUBE_PAM_MODULE_NAME")
```
Comment (liggitt) [inline plugin/pkg/auth/authenticator/request/basicauth/basicauth.go]: enabling/specifying the PAM module would be an API server flag, the presence of which would add a PAM password authenticator to the auth chain

Comment (liggitt) [inline plugin/pkg/auth/authenticator/request/basicauth/basicauth.go]: See
cmd/kube-apiserver/app/options/options.go
cmd/kube-apiserver/app/server.go
pkg/apiserver/authenticator/authn.go

Comment (arimed) [inline plugin/pkg/auth/authenticator/request/basicauth/basicauth.go]: ok, but to collect the user/password via the browser, you would still need basic auth turned on right?

Comment (liggitt) [inline plugin/pkg/auth/authenticator/request/basicauth/basicauth.go]: In general, basic auth from browsers against APIs is a really bad idea. it opens users up to CSRF, which means you need to add CSRF protection in front of the API, which means non-browser clients have to deal with CSRF protection. This plumbing will recognize credentials sent using basic auth, but it won't send basic auth challenges, so a browser wouldn't be prompted to enter a username/password.

Comment (arimed) [inline plugin/pkg/auth/authenticator/request/basicauth/basicauth.go]: agreed.  Note that in the current implementation (without my changes), if basic auth option is turned on for api server, then the browser does prompt the user for username & password.

Comment (liggitt) [inline plugin/pkg/auth/authenticator/request/basicauth/basicauth.go]: That is currently only for the toy basic auth file implementation. We should not extend that behavior to real password impls like this.

Comment (liggitt) [inline plugin/pkg/auth/authenticator/request/basicauth/basicauth.go]: See threads in https://github.com/kubernetes/kubernetes/pull/6508#issuecomment-90622897

File: plugin/pkg/auth/authenticator/request/basicauth/basicauth.go [hunk 6]
```diff
@@ -17,10 +17,9 @@ limitations under the License.
 package basicauth
 
 import (
-	"net/http"
```
Comment (liggitt) [inline plugin/pkg/auth/authenticator/request/basicauth/basicauth.go]: Undo this change

Comment (arimed) [inline plugin/pkg/auth/authenticator/request/basicauth/basicauth.go]: will do, one sec

File: pkg/apiserver/authenticator/authn.go [hunk 7]
```diff
@@ -59,6 +61,14 @@ func New(config AuthenticatorConfig) (authenticator.Request, error) {
 		authenticators = append(authenticators, basicAuth)
 	}
 
+	if config.PamPassword {
+		basicAuth, err := newAuthenticatorFromPamPassword(config.BasicAuthFile)
```
Comment (liggitt) [inline pkg/apiserver/authenticator/authn.go]: Should be PamPassword, right?

File: pkg/apiserver/authenticator/authn.go [hunk 8]
```diff
@@ -44,6 +45,7 @@ type AuthenticatorConfig struct {
 	ServiceAccountLookup      bool
 	ServiceAccountTokenGetter serviceaccount.ServiceAccountTokenGetter
 	KeystoneURL               string
+	PamPassword               bool
```
Comment (liggitt) [inline pkg/apiserver/authenticator/authn.go]: Make this a string that holds the name of the service, not a bool

File: cmd/kube-apiserver/app/server.go [hunk 9]
```diff
@@ -374,7 +375,7 @@ func Run(s *options.APIServer) error {
 			ReadWritePort:             s.SecurePort,
 			PublicAddress:             s.AdvertiseAddress,
 			Authenticator:             authenticator,
-			SupportsBasicAuth:         len(s.BasicAuthFile) > 0,
+			SupportsBasicAuth:         len(s.BasicAuthFile) > 0 || s.PamPassword,
```
Comment (liggitt) [inline cmd/kube-apiserver/app/server.go]: This flag controls sending basic auth challenges, which will prompt browsers for  passwords, without any way to log out, and without any cross site request forgery protection. We should not enable this flag for more password authentication methods without addressing those issues

File: cmd/kube-apiserver/app/options/options.go [hunk 10]
```diff
@@ -96,6 +96,7 @@ type APIServer struct {
 	SSHKeyfile                 string
 	MaxConnectionBytesPerSec   int64
 	KubernetesServiceNodePort  int
+	PamPassword                bool
```
Comment (liggitt) [inline cmd/kube-apiserver/app/options/options.go]: Also a string that holds the PAM service name

File: hack/lib/golang.sh (hunk 2 above)
Comment (liggitt) [inline hack/lib/golang.sh]: Is the lib cross platform? If not, you'll need to separate out the bits that are platform dependent (or call a platform dependent package) into a separate file, and use build tags to include them selectively, then create a separate file the implements the same functions and use build tags to select that file for all the unsupported platforms.

File: plugin/pkg/auth/authenticator/password/pampassword/pampassword.go [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+	switch s {
+	case pam.PromptEchoOn:
+		return c.User, nil
+	case pam.PromptEchoOff:
+		return c.Password, nil
+	}
+	return "", errors.New("unexpected")
+}
+
+// NewAllow returns a password authenticator that allows any non-empty username
```
Comment (liggitt) [inline plugin/pkg/auth/authenticator/password/pampassword/pampassword.go]: Fix godoc

Comment (liggitt) [inline plugin/pkg/auth/authenticator/password/pampassword/pampassword.go]: Take the PAM module name here and persist it in the struct, and use it instead of reading an envvar in AuthenticatePassword

File: plugin/pkg/auth/authenticator/password/pampassword/pampassword.go [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+// AuthenticatePassword implements authenticator.Password to allow any non-empty username,
+// using the specified username as the name and UID
+func (pamPaswdAuthenticator) AuthenticatePassword(username, password string) (user.Info, bool, error) {
+	c := Credentials{
+		Password: password,
+	}
+
+	module := os.Getenv("KUBE_PAM_MODULE_NAME")
```
Comment (liggitt) [inline plugin/pkg/auth/authenticator/password/pampassword/pampassword.go]: Read the module from the struct. I don't know enough about PAM to know if a default like this is expected or not, but I'd rather see it be explicitly specified

File: plugin/pkg/auth/authenticator/password/pampassword/pampassword.go [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+		return c.User, nil
+	case pam.PromptEchoOff:
+		return c.Password, nil
+	}
+	return "", errors.New("unexpected")
+}
+
+// NewAllow returns a password authenticator that allows any non-empty username
+func NewPamPassword() authenticator.Password {
+	return pamPaswdAuthenticator{}
```
Comment (liggitt) [inline plugin/pkg/auth/authenticator/password/pampassword/pampassword.go]: This is the prime file for splitting by platform. Have a pampassword_supported.go that contains this impl, and a pampassword_unsupported.go that contains a NewPAMPassword() func that returns `nil, errors.New("PAM password authentication is not supported")`
